*/etcd: enable tls for statically provisioned etcd

s-urbaniak opened the pull request:
Fixes #459

TODOS:
- [x] AWS (tested tls enabled/disabled, experimental enabled/disabled)
- [x] Baremetal (tested tls enabled/disabled, experimental enabled/disabled)
- [x] OpenStack
- [x] Test with experimental=true
- [x] configure additional Jenkins job
- [ ] VMWare (/cc @alekssaul) (out of scope)
- [ ] Azure (out of scope)

s-urbaniak commented:
That was a tough fight with terraform, now (finally) stuck at the API server. Maybe I'm blind after working on this, but the API server get's 504 gateway timeouts, while trying to persist objects. Debug output with `--v=7`:

```
[132.909µs] [8.792µs] About to store object in database                                                                             
"Create /api/v1/nodes" [30.00055836s] [30.000425451s] END                                                                           
I0518 17:07:48.976654       7 wrap.go:75] POST /api/v1/nodes: (30.00092623s) 504                                                    
goroutine 1180 [running]:                                                                                                           
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc420936f50, 0x1f8)                        
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:219 
+0xbb                                                                                                                               
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc420936f50, 0x1f8)                         
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/log.go:198 
+0x35                                                                                                                               
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*responseWriterDelegator).WriteHeader(0xc422ce23c0, 0x1f8)                 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:134 +0x45                                                                                                                        k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x6cb1fa0, 0xc420da5ec0, 0x0, 0x0, 0x40e913e, 0x2, 0x6ca9920, 0xc42267eb40, 0xc422c4a690, 0x1f8, ...)                                                                    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:89 +0x162                                                                                                       k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x6c7fe20, 0xc422dc4380, 0x6cb1fa0, 0xc420da5ec0, 0x0, 0x0, 0x40e913e, 0x2, 0x6ca9920, 0xc42267eb40, ...)                                                                      /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:106 +0x103                                                                                                      k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(0xc4212be280, 0x6c7fe20, 0xc422dc4380, 0x6ca9920, 0xc42267eb40, 0xc422c4a690)                                                                                                              /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:100 +0xd7                                                                                                                          
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.createHandler.func1(0xc422ce2330, 0xc422cf4060)                            /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:435 +0xee2                                                                                                                         k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc422ce2330, 0xc422cf4060)               
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:104 +0x1ec                                                                                                                       
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc420ef3ef0, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)  
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:272 +0xba0                                                                                                                             
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).(k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                                                       
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:120 +0x48                                                                                                                              
net/http.HandlerFunc.ServeHTTP(0xc42150e050, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                 
        /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             
net/http.(*ServeMux).ServeHTTP(0xc420fb60f0, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                 
        /usr/local/go/src/net/http/server.go:2022 +0x7f                                                                             
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x6ca99a0, 0xc420936f50, 0xc422c4a690)      
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:50 +0x365                                                                                                                  
net/http.HandlerFunc.ServeHTTP(0xc420fd7340, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                 
        /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x6ca99a0, 0xc420936f50, 0xc422c4a690)      
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:47 +0x1e4d                                                                                                                 
net/http.HandlerFunc.ServeHTTP(0xc420fd7380, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                         /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x6ca99a0, 0xc420936f50, 0xc422c4a690)             /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:73 +0x2c2                                                                                                                 net/http.HandlerFunc.ServeHTTP(0xc420ffb450, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                 
        /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.WithRequestContext.func1(0x6ca99a0, 0xc420936f50, 0xc422c4a690)     
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:107 +0xef                                                                                                                 
net/http.HandlerFunc.ServeHTTP(0xc420fa9840, 0x6ca99a0, 0xc420936f50, 0xc422c4a690)                                                         /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPanicRecovery.func1(0x6ca99a0, 0xc420936f50, 0xc422c4a690)         
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:74 +0x24a                                                                                                                              
net/http.HandlerFunc.ServeHTTP(0xc420fb71d0, 0x7f9770c62070, 0xc42267eb30, 0xc422c4a690)                                            
        /usr/local/go/src/net/http/server.go:1726 +0x44                                                                             k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc420fa9900, 0x6cb4ee0, 0xc42267eb30, 0xc422c4a690, 0xc422cf4000)                                                                                                       
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:91 +0x8d                                                                                                                            
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP                                         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:93 +0x1d1                                                                                                                                                                                                                                                               
logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12`\n\x04\n\x00\x12\x00\x12\aFailure\x1a9Timeout: request did not complete within allowed duration\"\aTimeout*\b\n\x00\x12\x00\x1a\x00(\x000\xf8\x03\x1a\x00\"\x00"                                           
I0518 17:07:54.758053       7 round_trippers.go:420] Response Status: 504 Gateway Timeout in 60000 milliseconds           [776/1821]
```

s-urbaniak commented:
A debug side-note to myself: a manual `curl -k https://${var.cluster_name}-etcd-1.${var.base_domain}` from the master vm succeeds.

s-urbaniak commented:
After digging through the k8s sources, it is quite apparent, why this must fail (at least for etcd3). This PR only supplies the self-generated etcd CA-file to the api-server. While this was perfectly fine with etcd2 [1], the k8s etcd3 client config requires all three parameters being set, see [2].

It is unfortunate that the api-server won't give any hint about this, but simply sets the TLS config to `nil` if not all three parameters are provided.

Short question towards @xiang90: Is upstream correct? Are the CA cert, keyFile, as well as the certificate really needed when `--client-cert-auth=false` when using grpc?

/cc @alexsomesan 

[1] https://github.com/kubernetes/kubernetes/blob/v1.6.4/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd2.go#L59-L80
[2] https://github.com/kubernetes/kubernetes/blob/v1.6.4/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L41-L43

xiang90 commented:
>  Is upstream correct?

It does not support self cert at the moment. 

But why do we want to use self generated cert? and why we cannot provide the CA even if it is self signed?

xiang90 commented:
Also for self hosted etcd, we need to follow https://github.com/coreos/etcd-operator/blob/master/doc/user/cluster_tls.md to enable TLS.

s-urbaniak commented:
@xiang90 thanks a lot for the pointer to the documentation! A quick answer to your question:

> > Is upstream correct?

> It does not support self cert at the moment.

> But why do we want to use self generated cert? and why we cannot provide the CA even if it is self signed?

I provide the CA in the case of self signed in this PR. Actually, that is the only thing I am providing to the api server omitting any client certs as client authentication is out of scope of this PR. I was just wondering why the api server does not accept a CA only but insists on having also client certs.

xiang90 commented:
>  I was just wondering why the api server does not accept a CA only but insists on having also client certs.

Oh. Yea. That behavior is not good. Probably worth fixing on upstream k8s.

s-urbaniak commented:
Ha, my bad ;-) The expression `len(c.CertFile) == 0 && len(c.KeyFile) == 0 && len(c.CAFile) == 0` mislead me, it evaluates to `false` if only CAFile is given ;-)

The actual error was somewhere completely else, namely the dns_names which were configured with the etcd certificate request.

Tested successfully using 1 and 3 cluster nodes, now testing in conjunction with experimental.

s-urbaniak commented:
@sym3tri @robszumski just as a FYI: me and @alexsomesan initially agreed on provisioning the generated etcd certs via ignition rather than via S3/scp. We know this is not ideal but doesn't require us to download/upload assets from withing the etcd nodes causing less complexity in the terraform templates. Pleas assert from the product side, whether this is a showstopper.

dghubble commented on platforms/metal/matchers.tf:
```diff
@@ -30,13 +30,27 @@ resource "matchbox_group" "controller" {
+    etcd_tls_enabled    = "${var.tectonic_etcd_tls_enabled}"
+    etcd_tls_ca_crt_pem = "${base64encode(module.bootkube.etcd_ca_crt_pem)}"
+    etcd_tls_crt_pem    = "${base64encode(module.bootkube.etcd_crt_pem)}"
+    etcd_tls_key_pem    = "${base64encode(module.bootkube.etcd_key_pem)}"
```
See above "DO NOT PLACE SECRETS IN USER-DATA"

s-urbaniak replied:
Agreed, this is still WIP, but thanks a lot for pointing out the criticality in the context of baremetal.

I believe in the case of baremetal the distribution of those can be accomplished the same way as with the kube secrets, namely scp'ing them to the master nodes, being the distribution method I was thinking about.

s-urbaniak commented:
@dghubble PTAL for the baremetal part. I am scp'ing the assets now the same way as kubeconfig and other assets.

s-urbaniak commented:
Seeing the same failure as https://github.com/coreos/tectonic-installer/pull/935#issuecomment-305206098, hence this must be unrelated:

```
15:06:29 1 error(s) occurred:
15:06:29 
15:06:29 * module.vpc.aws_internet_gateway.igw (destroy): 1 error(s) occurred:
15:06:29 
15:06:29 * aws_internet_gateway.igw: Error waiting for internet gateway (igw-748c5d13) to detach: couldn't find resource (31 retries)
15:06:29 
```

s-urbaniak commented:
ok to test

dghubble commented on platforms/metal/tectonic.tf:
```diff
@@ -18,6 +18,8 @@ module "bootkube" {
   service_cidr = "${var.tectonic_service_cidr}"
   cluster_cidr = "${var.tectonic_cluster_cidr}"
+  cluster_name = "${var.tectonic_cluster_name}"
+  base_domain  = "${var.tectonic_base_domain}"
```
These variables shouldn't be required for bare-metal

s-urbaniak replied:
ah, right, good catch!

dghubble commented on examples/terraform.tfvars.metal:
```diff
@@ -76,6 +76,11 @@ tectonic_etcd_count = "0"
+// (optional) If set to `true`, TLS secure communication for etcd will be used.
+// 
+// Note: This is variable has no effect if `tectonic_experimental` is set to `true`.
+// tectonic_etcd_tls_enabled = false
```
What about always enabling TLS when self-hosted etcd is in use? I don't see a case where a user would opt not to use TLS and this adds another confusing variable.

s-urbaniak replied:
You have a good point. /cc @sym3tri @robszumski whether we want to make this configurable at all.

PS: This PR does not cover self-hosted etcd (using the operator in experimental) but the dedicated self-provisioned etcd use case (non-experimental).

dghubble replied:
Thanks for the clarification that this PR is for on-host etcd only.

dghubble commented on examples/terraform.tfvars.metal:
```diff
@@ -50,7 +50,7 @@ tectonic_cluster_name = ""
 // Note: This works only when used in conjunction with an external etcd cluster.
 // If set, the variables `tectonic_etcd_servers`, `tectonic_etcd_client_cert_path`, and `tectonic_etcd_client_key_path` must also be set.
-// tectonic_etcd_ca_cert_path = ""
+// tectonic_etcd_ca_cert_path = "/dev/null"
```
Can you consider accepting the ca_cert rather than the path, which assumes a file is read from disk? Looking through the other variables in use in Tectonic, many of the _path variables probably need to go that way.

dghubble replied:
Seems like the client (user writing terraform.tfvars, module users, or the GUI) would pick whether the use a "${file("path")}" or provide the contents directly.

alexsomesan replied:
Putting materials like certs inline in the tfvars is not a direction we want to go in.

We have to account for the fact that users, in real production environments, may want to use specialized tools to manage distribution of secrets (for example Vault + VaultFS).
I can see how this issue may not be apartment during development, but we should be mindful about taking this flexibility away from production users.

/cc: @coresolve might have more real-world input on this topic

dghubble replied:
For real world usage, there needs to be a broader discussion about how Tectonic's generated assets can be cleanly split into non-secret infrastructure declarations (which get checked into infra repos) and secret generated files (which go into a secret store). For concerns like that, you may want to consider how the pure bootkube module does this by generating secrets into a target location (which works well for secret storage systems). https://github.com/coreos/matchbox/blob/master/examples/terraform/modules/bootkube/variables.tf#L60. This powers several colocations using a declarative repo and separate secret store.

A separate problem and followup is how to input user-define certificates (which come from a secret store) as variables.  I agree its a good idea to pull from a path, but I think that's done in the calling config, not in the module (context: this stuff is about to move to be a module).

Either way, I can patch this up when switching bare-metal over to be a module.

dghubble replied:
On second though, the distinction btw doing it in the module vs at the top level config will become much clearer once bare-metal is a module. Or chat with me in person. Feel free to use a path for now.

s-urbaniak replied:
Let's move this (very good!) discussion over to https://github.com/coreos/tectonic-installer/issues/710, being the umbrella issue for these questions.

s-urbaniak replied:
To add my 5 cents to the concrete implementation detail for this PR: In my very first iteration I indeed generated the certs as files in the bootkube module, but I was hitting the problem though that Terraform did not give me enough guarantees that this file was being fully generated before subsequent modules ran. I ran into race conditions constantly.

That was the main motivation to pass the cert content as internal TF state between modules which did not confuse TFs internal dependency resolution any more.

dghubble commented on platforms/metal/remote.tf:
```diff
@@ -1,5 +1,41 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /etc/ssl/etcd",
+      "sudo mv /home/core/etcd_ca.crt /etc/ssl/etcd/ca.crt",
```
Here you copy files to /etc/ssl/etcd, but in the Container Linux config, a different path is used (e.g. `ETCD_CERT_FILE=/etc/ssl/certs/client.crt`). Am I missing something?

dghubble replied:
Ah, nvm. Inside the etcd-wrapper script, ETCD_SSL_DIR is read to control the volume mounting into /etc/ssl/certs. Looks good.

diegs commented on modules/aws/etcd/ignition.tf:
```diff
@@ -48,14 +81,19 @@ data "ignition_systemd_unit" "etcd3" {
-  --listen-peer-urls=http://0.0.0.0:2380
+  --advertise-client-urls=${var.tls_enabled ? "https" : "http"}://${var.cluster_name}-etcd-${count.index}.${var.base_domain}:2379 \
+  ${var.tls_enabled
+      ? "--cert-file=/etc/ssl/etcd/client.crt --key-file=/etc/ssl/etcd/client.key --peer-cert-file=/etc/ssl/etcd/client.crt --peer-key-file=/etc/ssl/etcd/client.key --peer-trusted-ca-file=/etc/ssl/etcd/ca.crt -peer-client-cert-auth=true"
```
Is there a reason for using the same certs for peer and client communication? In general I believe it's recommended that you use separate ones, and I don't see a strong reason not to.

cc @xiang90

xiang90 replied:
I talked to @dghubble about this exact thing today. @dghubble wants to reuse the certs for easier management. The best way is to separate them and also generate client certs from the trusted CA for other component consumption.

xiang90 replied:
@dghubble Or maybe a different thing? It is not the bare metal one after my second look.

dghubble replied:
I answered below, that was a different thing for https://github.com/coreos/matchbox/pull/566

dghubble commented:
I was talking about file permissions for certs for locksmith and etcd, for
the matchbox repo, different topic.

Here, I'd say using the separate peer and client certs is a good idea.

On Jun 6, 2017 8:23 PM, "Xiang Li" <notifications@github.com> wrote:

> *@xiang90* commented on this pull request.
> ------------------------------
>
> In modules/aws/etcd/ignition.tf
> <https://github.com/coreos/tectonic-installer/pull/751#discussion_r120527113>
> :
>
> >  ExecStart=
>  ExecStart=/usr/lib/coreos/etcd-wrapper \
>    --name=etcd \
>    --discovery-srv=${var.base_domain} \
> -  --advertise-client-urls=http://${var.cluster_name}-etcd-${count.index}.${var.base_domain}:2379 \
> -  --initial-advertise-peer-urls=http://${var.cluster_name}-etcd-${count.index}.${var.base_domain}:2380 \
> -  --listen-client-urls=http://0.0.0.0:2379 \
> -  --listen-peer-urls=http://0.0.0.0:2380
> +  --advertise-client-urls=${var.tls_enabled ? "https" : "http"}://${var.cluster_name}-etcd-${count.index}.${var.base_domain}:2379 \
> +  ${var.tls_enabled
> +      ? "--cert-file=/etc/ssl/etcd/client.crt --key-file=/etc/ssl/etcd/client.key --peer-cert-file=/etc/ssl/etcd/client.crt --peer-key-file=/etc/ssl/etcd/client.key --peer-trusted-ca-file=/etc/ssl/etcd/ca.crt -peer-client-cert-auth=true"
>
> @dghubble <https://github.com/dghubble> Or maybe a different thing? It is
> not the bare metal one after my second look.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/tectonic-installer/pull/751#discussion_r120527113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACJidELfPoT1YisfYmkJ-LUE7GpyjXkUks5sBhe7gaJpZM4NfaGH>
> .
>

s-urbaniak commented:
@dghubble @xiang90 ack, I will include separate peer and client cert generation in this PR.

dghubble commented:
@s-urbaniak Enabling on-host TLS was added upstream in the Matchbox bare-metal reference clusters by https://github.com/coreos/matchbox/pull/566. The changes end up being similar so maybe helpful. You might be able to just pickup the etcd peer cert/key change and we also noticed etcd's TLS docs request certs have owner:group `etcd:etcd` and limited permissions.

I still wonder if some of the complexity in this PR could be reduced if on-host (static as you say) etcd was *always* deployed with TLS. I was able to use that strategy to just gate on one variable `experimental_self_hosted_etcd = true/false`. And once we can enable TLS for self-hosted etcd, some additional conditions go away.

s-urbaniak commented:
@dghubble thanks a lot for the pointer! I agree with your point for the potential to reduce code complexity if we just enable tls by default, hence re-pinging @robszumski, @sym3tri, @Quentin-M if we can just skip `tectonic_etcd_tls_enabled` and apply TLS by default.

alexsomesan commented:
@dghubble @s-urbaniak I also support the idea of always configuring TLS for on-host etcd. We can definitely benefit from the simplified configuration this approach allows for.

Can't really see a use-case for not having TLS configured, other than temporary debugging situations. In that case I think it's worth the effort to disable by hand if needed.

diegs commented:
Sorry, I'm late to the party here, but this is for external etcd right? Does that encompass people who are bringing their own etcd clusters? If so, technically we can't force them to use TLS (though obviously they should).

Or is this only for a subset of that scenario, where we deploy an external (non-self-hosted) etcd cluster for them?

Either way, if this is a behavior change from before we may need to think about the impact on upgrades. It might not be anything, but please let me know if my understanding above is incorrect, and then we can investigate how these changes might affect our expectations in the future of what etcd configuration in a cluster looks like.

s-urbaniak commented:
@diegs  please see my comments below
> Sorry, I'm late to the party here, but this is for external etcd right? Does that encompass people who are bringing their own etcd clusters? If so, technically we can't force them to use TLS (though obviously they should).
> Or is this only for a subset of that scenario, where we deploy an external (non-self-hosted) etcd cluster for them?

You actually have a good point. If we remove `tectonic_etcd_tls_enabled` from this PR and default to TLS, then people using non-TLS bring-your-own etcd clusters will be locked out. They will need to upgrade to TLS. That's a fair counter-argument.

> Either way, if this is a behavior change from before we may need to think about the impact on upgrades. It might not be anything, but please let me know if my understanding above is incorrect, and then we can investigate how these changes might affect our expectations in the future of what etcd configuration in a cluster looks like.

Your understanding actually is correct.

alekssaul commented on modules/openstack/etcd/ignition.tf:
```diff
@@ -27,6 +88,7 @@ data "ignition_file" "resolv_conf" {
 }
 
 data "ignition_systemd_unit" "etcd3" {
+  count  = "${var.instance_count}"
```
I believe this will require ignition_config resource to change from
```
  systemd = [
    "${data.ignition_systemd_unit.locksmithd.id}",
   ...
```
to 
```
  systemd = [
    "${data.ignition_systemd_unit.locksmithd.*.id[count.index]}",
   ...
```

alekssaul commented on modules/openstack/etcd/ignition.tf:
```diff
@@ -37,27 +99,42 @@ data "ignition_systemd_unit" "etcd3" {
 }
 
 data "ignition_systemd_unit" "locksmithd" {
+  count = "${var.instance_count}"
```
I believe this will require ignition_config resource to change from
```
  systemd = [
    "${data.ignition_systemd_unit.locksmithd.id}",
   ...
```
to 
```
  systemd = [
    "${data.ignition_systemd_unit.locksmithd.*.id[count.index]}",
   ...
```

s-urbaniak replied:
good catch, that change is necessary indeed, thanks!

s-urbaniak commented on Jenkinsfile:
```diff
@@ -139,6 +155,20 @@ pipeline {
               }
             }
+          },
+          "Terraform: Bare Metal etcd-tls": {
```
@Quentin-M question: is this safe given the following comment? https://github.com/coreos/tectonic-installer/blob/47f4f81/tests/smoke/bare-metal/smoke.sh#L11-L13

I am unsure about the invariants guaranteed by the Jenkins executor. The PR is green, so it worked, most probably because those bare-metal tests have been executed on different nodes. But does Jenkins always ensure this invariant for parallel tests?

Quentin-M replied:
All good :)

Quentin-M replied:
Is there a reason we would test this on bare-metal as well?

s-urbaniak commented:
This is ready for final review. The main decision point to take is whether to enable TLS by default with the cost of locking existing non-TLS bring-your-own etcd users.

s-urbaniak commented:
ok to test

s-urbaniak commented:
@lblackstone @coreypobrien feel free to give this PR a try on OpenStack as it introduces the possibility to have TLS backed communication to etcd.

coreypobrien commented on modules/openstack/etcd/ignition.tf:
```diff
@@ -1,20 +1,81 @@
   systemd = [
-    "${data.ignition_systemd_unit.locksmithd.id}",
+    "${data.ignition_systemd_unit.locksmithd.*.id[count.index]}",
     "${data.ignition_systemd_unit.etcd3.id}",
```
We'll need this to be `"${data.ignition_systemd_unit.etcd3.*.id[count.index]}"`

s-urbaniak replied:
Ah, good to know, thanks for the hint!

Quentin-M commented:
Thanks for your work on this!

@s-urbaniak Just talked with Rob. We don't clearly see a reason we should keep support for non-TLS etcd setups. BYO-etcd without TLS sounds scary, at best, and we should not encourage it. They'd still be able to do it if really necessary by modifying the Terraform sources. Does that sound fair?

As long as it does not break updates (?), let's do it. It will help keeping the code simple too. To be on the safe side, should still keep a copy of your branch, in case some other product input rises.

discordianfish commented:
Will this get merged without azure and vmware support? I can work on the azure bits next week but would much appriciate it if either this here or #999 get merged, so I don't need to jungle all the branches. :)

alekssaul commented:
I've worked on support for this on VMware and have a working copy. Based on my discussions with the team my intentions is to wait for this to merge in it's final form before submitting a follow up PR for VMware platform.

discordianfish commented:
Makes sense, can do the same for azure unless someone already has code for that.

Quentin-M commented:
PR on Bootkube for TLS self-hosted etcd: https://github.com/kubernetes-incubator/bootkube/pull/592

Quentin-M commented:
Hey,

We should remove support for non-TLS deployments as mentioned in https://github.com/coreos/tectonic-installer/pull/751#issuecomment-308191172, re-test and merge this. The PR is already quite large - support for the alpha platforms can be completed separately.

s-urbaniak commented:
@Quentin-M @alexsomesan As discussed yesterday OOB I rebased the PR leaving the option to use TLS for etcd and defaulting it to `true`. Tests are green, I switched the default AWS and bare-metal smoke tests to use TLS rather than redundantly adding new ones, as that is the direction we are heading anyways. PTAL

s-urbaniak commented:
@Quentin-M Note though, that Azure is not yet supported in this PR. I can follow-up a PR, once Azure@master is stable enough (/cc @alexsomesan)
